### PR TITLE
fix(torch): support compiled matrix-to-COO conversion

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Make exact matrix-to-COO conversion compatible with
`torch.compile(fullgraph=True)` when the edge count changes between calls.
The conversion now uses one `nonzero` compaction followed by indexed gathers,
while preserving eager output and overflow behavior.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

N/A

## Changes Made

- Replace three dynamic mask operations with one row-major `nonzero` operation
  and indexed gathers for neighbor indices and shifts. Exact allocation remains
  data-dependent and may synchronize the host.
- Preserve exact COO shapes, ordering, pointers, shift alignment, empty inputs,
  and eager `NeighborOverflowError` fields. Compiled overflow uses an
  asynchronous assertion.
- Require PyTorch >=2.10 for Torch extras and document compiled conversion and
  overflow behavior.
- Extend existing public-path tests for conversion parity, edge cases, changing
  compiled edge counts, and eager and compiled overflow.

## Testing

- [ ] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

Validation completed:

- `make test-minimal` passed with 74% coverage.
- Affected Torch neighbor suites: 1,261 passed and 139 skipped.
- Focused compiled CPU and CUDA tests: 24 passed.
- PyTorch 2.10 and the locked PyTorch 2.12 environment passed changing-edge-count
  fullgraph compilation on CPU and CUDA.
- `make lint`, `make interrogate`, `make license`, `uv lock --check`,
  documentation build, and `git diff --check` passed.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [ ] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

PyTorch 2.8 and 2.9 cannot compile the default fullgraph dynamic-output path;
PyTorch 2.10 succeeds, motivating the Torch extras dependency floor.

Matrix-to-COO conversion was benchmarked on periodic float32 CsCl systems using
an NVIDIA RTX 6000 Ada GPU.

| Matrix rows x capacity | Edges | Baseline (ms) | This PR (ms) | Change |
| ---: | ---: | ---: | ---: | ---: |
| 250 x 64 | 6,500 | 0.166 | 0.116 | -29.9% |
| 2,000 x 64 | 52,000 | 0.163 | 0.113 | -31.0% |
| 8,192 x 64 | 212,992 | 0.162 | 0.112 | -30.8% |
| 101,307 x 64 | 2,633,956 | 0.438 | 0.329 | -24.9% |

These results are workload-specific and do not imply a universal speedup.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/torch/neighbors/neighbor_utils.py [hunk 1]
```diff
@@ -444,7 +464,7 @@ def get_neighbor_list_from_neighbor_matrix(
     torch.cumsum(num_neighbors, dim=0, out=neighbor_ptr[1:])
 
     if neighbor_shift_matrix is not None:
-        neighbor_list_shifts = neighbor_shift_matrix[mask]
+        neighbor_list_shifts = neighbor_shift_matrix[i_idx, slot_idx]
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/neighbors/neighbor_utils.py:467]: Fixed in eebf3ff.

**Review by dallasfoster — APPROVED** [review]

**Review by dallasfoster — APPROVED** [review]
